### PR TITLE
Add clang-format configuration files and a helper script

### PR DESCRIPTION
**Pull request by dreamer**
Decided to move clang-format part out of contributing guidelines draft, as I consider this part fairly done. It's already very helpful, especially when configured with developer's editor - instructions for popular editors (Vim, Emacs, Visual Studio, some others) are here: https://releases.llvm.org/8.0.0/tools/clang/docs/ClangFormat.html

Two tiny differences from our convention are about formatting of constructor initializer list:
- when initializer list is short enough to fit after constructor declaration, formatter will use one line
- we used tab (+2 spaces for second and later line) for indentation of initializers, clang-format uses 8 spaces (+2 for second and later line). It looks the same, except formatting will be now consistent with our formatting of parameters in function declarations and function calls.

One bigger difference, that is not available in 8.0 (but will be available after upgrade to 9.0) is formatting of macro definitions - eventually, we'll want `AlignConsecutiveMacros: true` ([description](https://releases.llvm.org/9.0.0/tools/clang/docs/ClangFormatStyleOptions.html)).

Other details of what and why are in git commit messages, so I'll just paste them here:
```
commit cf6dbc3bc33cea2747818200276de1a4e68585bb
Author: Patryk Obara <dreamer.tan@gmail.com>
Date:   Mon Mar 30 06:47:11 2020 +0200

    Implement script for re-formatting recent commits
    
    This script is intended to be used in two situations:
    
    - When user can't configure clang-format in code editor or IDE
    - Potentially to be used in CI in the future
    
    Normal usecase is to run this script after commit, then review and
    add formatted code, and amend the commit, e.g.:
    
      git commit -m "Edit some C++ code"
      ./scripts/format-commit.sh
      git add -p  # select formatting changes you want to use
      git commit --amend
    
    Run:
    
      ./scripts/format-commit.sh --help
    
    to learn about other options.

commit 6fff29854351a5f71b4817abe242892c818f85c0
Author: Patryk Obara <dreamer.tan@gmail.com>
Date:   Sun Mar 29 05:55:18 2020 +0200

    Add clang-format rules
    
    Rules are defined for all C/C++ code, except src/libs.
    
    These rules are designed to emulate kernel coding convention with some
    adaptations for C++11. Configuration file uses options available in
    clang-format 8.0 (the newest version in 10.0) to provide compatibility
    with older environments and operating systems.  Some options are
    commented out and marked with TODOs - those need to be investigated and
    enabled/disabled during upgrade to clang-format 9.0.
    
    DOSBox codebase is a mix of various styles with only few consistent
    rules seen throughout the codebase, these clang-format rules preserve
    some of existing conventions and conciously break with others.
    
    Some unwritten upstream rules, that are now encoded in clang-format:
    
    - Using tab for indentation.
    - K&R-like indentation for control statements.
    - Indentation rules for structs, classes, and non-anonymous enums.
    - Case labels in switch statements are aligned the same way goto
      statements would be (also K&R rule).
    
    Some formatting aspects that were not followed consistently throughout
    old DOSBox code, but are now encoded in clang-format:
    
    - Space placing in control statements and function calls (makes the code
      much more readable).
    - Control statements (if-else, while) must be broken into multiple lines
      (makes the code more readable, helps with debugging, reading compiler
      logs and static analysis reports).
    
    Some unwritten upstream rules, that are now changed by these rules:
    
    - Placing opening function bracket in the same line as function
      declaration (not in line with K&R).  This rule makes it hard to make
      constructor formatting consistent - old code dealt with it by not
      formatting initializer lists at all.  Unformatted initializer list can
      result in lines hundreds of lines long and make it hard to add/remove
      class fields while assuring correct initialization order.
      In new clang-format rules K&R indentation is followed, allowing
      initalizer lists to be formatted one initializer per line (which makes
      it much easier to read and edit.
```

### CODE REVIEW

File: scripts/format-commit.sh [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+		| filter_line_range \
+		| to_clang_line_range
+}
+
+prepare_clang_params () {
+	local -r file=$1
+	local -r range=$(git_diff_to_clang_line_range "$file")
+	# print file name only when there are any lines detected, otherwise
+	# clang-format would process the whole file
+	if [ -n "$range" ]; then
```
Comment (kcgen) [inline scripts/format-commit.sh]: (minor) `[[` double brackets is handled by bash, and avoids running the `[` command.

File: scripts/format-commit.sh [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+	local -r file=$1
+	git_diff --ignore-space-at-eol -U0 HEAD~1 "$file" \
+		| grep -E "^@@" \
+		| filter_line_range \
+		| to_clang_line_range
+}
+
+prepare_clang_params () {
+	local -r file=$1
+	local -r range=$(git_diff_to_clang_line_range "$file")
```
Comment (kcgen) [inline scripts/format-commit.sh]: I'm surprise the bash-checker isn't complain about unquoted variables here?

File: scripts/format-commit.sh [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+show_tip () {
+	if [ -n "$(git_diff HEAD)" ] ; then
+		echo
+		echo "clang-format formatted some code for you."
+		echo
+		echo "Run 'git diff' to see what changed."
+		echo "Run 'git commit -a --amend' to save the result."
+		exit 0
+	fi
+}
```
Comment (kcgen) [inline scripts/format-commit.sh]: `assert_empty_diff`  and `show_tip` are pretty similar except for the return code .
- If assert_empty sees a diff, it prints a tip then exits 1.
- If show_tip sees a diff, it prints a tip then exits 0.  
Maybe room for combining? or break out into three functions:
 - assert_diff -> if exists, show tip and exit 0
 - assert_empty_diff -> if exists, show tip and exit 0
 - show_top -> print the body.

File: scripts/format-commit.sh [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+format () {
+	local -r since_ref=${1:-HEAD~1}
+	pushd "$(git rev-parse --show-toplevel)" > /dev/null
+	echo "Using paths relative to: $(pwd)"
+	find_cpp_files "$since_ref" | run_clang_format
+	popd > /dev/null
+}
+
+assert_empty_diff () {
+	if [ -n "$(git_diff HEAD)" ] ; then
```
Comment (kcgen) [inline scripts/format-commit.sh]: Prefer `[[` double brackets

File: .clang-format [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+PenaltyReturnTypeOnItsOwnLine: 200
+
+# Don't sort "#include" declarations.
+#
+# clang-format has very rich customization options for grouping and sorting
+# "include" directives, but there are still edge cases, so we still need to
+# rely on developers doing this manually.
+#
+# See https://github.com/dreamer/dosbox-staging/issues/196 for details.
+#
```
Comment (kcgen) [inline .clang-format]: Would be nice if we could have some logic here:
 - if <ctypes> and <algorithm> present, then don't sort.
 - otherwise re-arrange per X, Y, Z rules.

Likewise, re-arranging the overall blocks of headers (system headers before local headers) would be nice.

Sample with re-arranging just the local headers in `"` quotes might be doable, if there's a distinctions. 

All just nice-to-have, if it can do it.

**Comment by kcgen** [comment]
This will big a huge benefit for all contributors, and a time-saver for reviews given I would say up maybe 30% to 50% of your comments to new contributors (including myself) involved formatting. Congrats to getting this done!

Just a couple minor comments if you agree with them. Overall it looks great; but none are blockers.  

I'm curious about the "no-format" sub directories we have; maybe the reformat helper script can take in an exclusion pattern (arg or an env variable), or simply hardcode the directory patterns?  That said; I wouldn't mind reformat a handful of files in src/lib/decoders, atleast those files fully under our control.

File: .clang-format (hunk 5 above)
Comment (dreamer) [reply .clang-format:282]: Formatter needs to arrive at includes order deterministically, so include grouping rules depend on sorting :( There's an option to group and sort each group separately and define multiple groups though.

I have a few ideas on how to resolve it so we could enable it in the future, but it will need some experimentation and a separate PR.

File: scripts/format-commit.sh (hunk 3 above)
Comment (dreamer) [reply scripts/format-commit.sh:106]: Hmm, but descriptions are slightly different (they were a lot different during development). The only common thing in here is existence of `if` statement.

File: scripts/format-commit.sh (hunk 2 above)
Comment (dreamer) [reply scripts/format-commit.sh:149]: That's because these are assignments (R-value is automatically quoted in this case) - there's no way anything after `=` to accidentally merge with anything else in the line. Parameters in the subshell still needs to be quoted (`"$file"`).

**Comment by dreamer** [comment]
I'll just merge this in and work on bash improvements in a separate PR. I'm missing formatting capability on my other branches.

Thanks for the review! :) I had older version of this script laying around for some time and was seriously considering rewriting it from the scratch in python at last… but processing of git plumbing commands is initially easier in bash and turns out it was not a problem to decode my old intentions from old code.

**Comment by dreamer** [comment]
> I'm curious about the "no-format" sub directories we have; maybe the reformat helper script can take in an exclusion pattern (arg or an env variable), or simply hardcode the directory patterns? That said; I wouldn't mind reformat a handful of files in src/lib/decoders, atleast those files fully under our control.

That noformat can be overridden by placing another `.clang-format` file inside specific lib directory. It can't apply different styles to files in the same directory though.

As for whitelisting files to be fully-reformatted - I think it's too early for that. Let's learn how the current ruleset work for us in practice and if there are any undesirable results. One example that we need to keep an eye for is formatting of anonymous enums (but anonymous enums are kind-of silly workaround if you think about it and could be safely replaced by `constexpr` global variables or macro definitions).

I forgot to ask earlier: what version of `clang-format` do you have installed? My oldest OS has 8.0 - I want to balance new formatting tweaks available in newer versions with wider audience by keeping the older version.

**Comment by kcgen** [comment]
> I'll just merge this in and work on bash improvements in a separate PR. I'm missing formatting capability on my other branches.

Sounds good!

**Comment by dreamer** [comment]
> > I'll just merge this in and work on bash improvements in a separate PR. I'm missing formatting capability on my other branches.
> 
> Sounds good!
> I have clang-format 6 through 10, with 9 as the default.

Great :) Debian testing has 9.0 as well, Ubuntu 16.04 LTS stopped at 8.0, but 18.04 includes version 9 in the repos; Arch world obviously has it.

This month Ubuntu 20.04 will be released and Fedora 30 will be 1 month away from EOL, so I think we'll be able to upgrade rules to 9.0 in few weeks… Unless there will be some problem with clang-format plugin for VS (oops, I haven't actually tested it yet).

**Comment by kcgen** [comment]
8, moving to 9 soon, sounds reasonable.  I think more developers here will have access to extended repos that give them close to the latest compilers.  

I'm also sure that our use cases aren't bread and butter for what it handles, so hopefully were not missing out on much.

Do you know if the file is extensible in that future additions are simply ignored by older versions?

**Comment by dreamer** [comment]
> Do you know if the file is extensible in that future additions are simply ignored by older versions?

Unrecognized options cause errors; clang developers are very careful about adding new ones.

That's why I decided to stick to 8.0 for the moment. I was testing with 8.0 and 9.0 on Linux. Just tested VS clang-format plugin and it's using version 10.0, so we don't need to worry about Windows developers.

I have good feeling about this - hopefully it'll seriously limit formatting-related bikeshedding :)

**Comment by kcgen** [comment]
Ahh safe call on sticking with 8.
We could also add it as a verify script ane jnclude it in the linters?
Also thinking we could add it to the pre-commit hooks (in our local repos), so it helps us on every commit.
Noje the less, I'm looking forward to putting it to use! I will try to reformat my autotype PR with it and see what it does :-)

**Comment by dreamer** [comment]
clang-formatter is based on clang compiler, so we can't really bundle it with the repo… but it's not a problem, most people will have it available.

It really works best when integrated into the editor - in Vim I can precisely select the lines of code I want to be formatted, press Ctrl-K, and done :) Sometimes even just to make old code readable for a moment. In VS shortcut is Ctrl-R Ctrl-F if I recall correctly (although I use Vim also on Windows, I use VS only as compiler/debugger).

For me, integrating it into hooks would break my workflow (I would need to reload files in editor after a commit), but I guess some clever solution using git filters is possible (although I will stick to using it inside Vim).

**Comment by kcgen** [comment]
Just for fun I had it process `cdrom_image.cpp` and `cdrom.h` (probably the file that I've touched the most over these years).   I would say 95% of lines are great and overall is an improvement.  So the following comments are just the corner cases or areas where I have questions:

---
**Wrapped Argument Indenting**

LOG_MSG's arguments don't get indented inside the `#if 0`, but they do in other `#ifdef DEBUG`



I guess the formatting happens after preprocessing, as expected - so that would explain it.  But wanted to share it.

---
**Wrapped Logic Statements**
This block here containing `||`'s and `&&`s has become quite a mess:


I would prefer wrapped logic expressions are _prefixed_ with the bolean operator because it allows for natural alignment - and easier for the eyes to get the overall branching structure:

``` c++
if ( // a really long check
    some_really_long > check_check
    || another_smaller_check()
    && (isTiny()
        || another_really_long_check >= some_other_long_value
        || tiny()))
```

Versus when the boolean operators are at the back, there is no alignment and it's much harder to instantly see the organization of the branches. 

``` c++
if ( // a really long check
    some_really_long > check_check ||
    another_smaller_check() &&
    (isTiny() ||
     another_really_long_check >= some_other_long_value ||
     tiny()))
```

---
**Constructor initialization side-by-side**

(Recall you mentioning this; and just want to confirm that we would manually wrap these back down to one-variable-per-line?



---

**Define blocks**

It kind of mangled this section of defines, which admittedly use really long lines.
But some things jump our at me - the ragged and spaced-out `\`'s at the end.
In some cases the defined value was wrapped beneath the `#define`, so I'm not sure why that happened.  Maybe it treats the comment as part of the define (and won't relocate the comment?)



---

**Header function argument wrapping**

In `cdrom.h`, handling of the wrapped arguments was extremely inconsistent:





---

**Alignment fully lost in member blocks**




Header files are like the table of contents of a book, which big-picture structure take prescident over the content itself.  It's important for the paragraph numbers to be aligned (and sub-indented correctly), likewise the page number at the back should also be aligned).  So too should header files; it's really a reference for a human.  Because of this, I personally would like to see more structure and alignment in the headers. 

Then in the written code.. the body of the book, sure - everything can be sqaushed left and nothing really aligned because we naturally read left to right. (and I'm OK with that approach).

**Comment by dreamer** [comment]
I did use those files when developing and testing the ruleset, and in some cases I needed to do hard decisions when it really conflicts with style in other files :(

> **Wrapped Argument Indenting**

LOG_MSG got formatted for me differently?
```
	// Guard if we have no tracks or the sector is beyond the lead-out
	if (sector > MAX_REDBOOK_SECTOR || tracks.size() < MIN_REDBOOK_TRACKS ||
	    sector >= tracks.back().start) {
		LOG_MSG("CDROM: GetTrack at sector %u is outside the"
		        " playable range",
		        sector);
		return tracks.end();
	}
```
And after a tweak to the line:
```
		LOG_MSG("CDROM: GetTrack at sector %u is outside the playable  "
		        "range",
		        sector);
```
I tried to ramp up `PenaltyBreakString` to prevent it, but it does not seem to work the way I thought it will… :/

> **Wrapped Logic Statements**

This is one of reasons why I don't advocate for reformatting of whole files, commit hooks, or putting in in CI, I even used the following statement in config file:
```
# This style definition should only be understood as a hint
# for writing new code. The rules are still work-in-progress and do
# not yet exactly match the style we have in the existing code.
```
Yes, when logic statements are wrapped in tree-like structure like your example here, the results can be weird. Some options for breaking of logic and arithmetic expressions appeared in 10.0 (I think), but I don't think this exact formatting style can be accomplished.

However, there's something weird with your screenshot? For me it got formatted like this:
```
	// pvd[0] = descriptor type, pvd[1..5] = standard identifier,
	// pvd[6] = iso version (+8 for High Sierra)
	return ((pvd[0] == 1 && !strncmp((char *)(&pvd[1]), "CD001", 5) &&
	         pvd[6] == 1) ||
	        (pvd[8] == 1 && !strncmp((char *)(&pvd[9]), "CDROM", 5) &&
	         pvd[14] == 1));

```

> **Constructor initialization side-by-side**

Argh, I kept testing it over and over before settling on configuration, and this edge case still slipped in :/

I hope `AllowAllConstructorInitializersOnNextLine: false` will address it, but I need to test it. (option starting with in 9.0).

> **Define blocks**

See comment in conf file next to `AlignConsecutiveMacros` option. We will enable this in 9.0.
As for `\` - they for formatted for me correctly (they are lined up all on the same column).

> **Header function argument wrapping**

This is how it gets formatted for me: I see exactly the same formatting in vim and VS:
```
class CDROM_Interface {
public:
	virtual ~CDROM_Interface(void) {}
	virtual bool SetDevice(char *path) = 0;
	virtual bool GetUPC(unsigned char &attr, char *upc) = 0;
	virtual bool GetAudioTracks(uint8_t &stTrack, uint8_t &end, TMSF &leadOut) = 0;
	virtual bool GetAudioTrackInfo(uint8_t track, TMSF &start, unsigned char &attr) = 0;
	virtual bool GetAudioSub(unsigned char &attr,
	                         unsigned char &track,
	                         unsigned char &index,
	                         TMSF &relPos,
	                         TMSF &absPos) = 0;
	virtual bool GetAudioStatus(bool &playing, bool &pause) = 0;
	virtual bool GetMediaTrayStatus(bool &mediaPresent,
	                                bool &mediaChanged,
	                                bool &trayOpen) = 0;
	virtual bool PlayAudioSector(const uint32_t start, uint32_t len) = 0;
	virtual bool PauseAudio(bool resume) = 0;
	virtual bool StopAudio(void) = 0;
	virtual void ChannelControl(TCtrl ctrl) = 0;
	virtual bool ReadSectors(PhysPt buffer,
	                         const bool raw,
	                         const uint32_t sector,
	                         const uint16_t num) = 0;
	virtual bool LoadUnloadMedia(bool unload) = 0;
	virtual void InitNewMedia(void){};
};
```
> **Alignment fully lost in member blocks**

There several tweaks affecting declaration blocks or assignments blocks and I made concious decision to go with this formatting (after long deliberation). I would need to write separate long post about it… Overall - if you think we should change it (this specific thing), let's do it in a new PR, but any single configuration I tried resulted in surprising results *somewhere*, so I went with least disruptive one.

Options affecting this are: `AlignConsecutiveAssignments`, `AlignConsecutiveDeclarations`, `AlwaysBreakBeforeMultilineStrings`, `PenaltyBreakAssignment`; maybe you'll fine some option I missed in https://releases.llvm.org/8.0.0/tools/clang/docs/ClangFormatStyleOptions.html#
